PR: https://github.com/odoo/odoo/pull/236480

From: ec66f778dc7532fe9d41b051820052ac7827bbe5
From: Gurpreet Singh
Date: 2026-02-17 03:49:47

Breaking data model changes scores: del:2 + add:0, change matches:
-    notification_moment = fields.Selection([
-    moment = fields.Selection([

Total Changes: 45

[IMP] lunch: change time format from 12-hour to 24-hour

Before this commit, the lunch module automatic_email_time and notification_time
time field used a 12-hour format with AM/PM.

This update changes it to a 24-hour format to avoid confusion and follow a
standard time format.

Impact:
Users now see time in 24-hour format without AM/PM.

task-5270259

closes odoo/odoo#236480

Related: odoo/upgrade#8892
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/lunch/models/lunch_alert.py
+++ b/addons/lunch/models/lunch_alert.py
@@ -7,13 +7,13 @@ from zoneinfo import ZoneInfo
 from odoo import api, fields, models, _
 from odoo.fields import Domain
 
-from .lunch_supplier import float_to_time
+from odoo.tools.date_utils import float_to_time
 
 from odoo.addons.base.models.res_partner import _tz_get
 
 _logger = logging.getLogger(__name__)
 WEEKDAY_TO_NAME = ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun']
-CRON_DEPENDS = {'name', 'active', 'mode', 'until', 'notification_time', 'notification_moment', 'tz'}
+CRON_DEPENDS = {'name', 'active', 'mode', 'until', 'notification_time', 'tz'}
 
 
 class LunchAlert(models.Model):
@@ -35,9 +35,6 @@ class LunchAlert(models.Model):
         ('last_month', 'Employee who ordered last month'),
         ('last_year', 'Employee who ordered last year')], string='Recipients', default='everyone')
     notification_time = fields.Float(default=10.0, string='Notification Time')
-    notification_moment = fields.Selection([
-        ('am', 'AM'),
-        ('pm', 'PM')], default='am', required=True)
     tz = fields.Selection(_tz_get, string='Timezone', required=True, default=lambda self: self.env.user.tz or 'UTC')
     cron_id = fields.Many2one('ir.cron', ondelete='cascade', required=True, readonly=True)
 
@@ -58,8 +55,8 @@ class LunchAlert(models.Model):
     location_ids = fields.Many2many('lunch.location', string='Location')
 
     _notification_time_range = models.Constraint(
-        'CHECK(notification_time >= 0 and notification_time <= 12)',
-        'Notification time must be between 0 and 12',
+        'CHECK(notification_time >= 0 and notification_time <= 24)',
+        'Notification time must be between 0 and 24',
     )
 
     @api.depends('mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun')
@@ -92,7 +89,7 @@ class LunchAlert(models.Model):
                 and (not alert.until or fields.Date.context_today(alert) <= alert.until)
             )
 
-            sendat_tz = datetime.combine(fields.Date.context_today(alert, fields.Datetime.now()), float_to_time(alert.notification_time, alert.notification_moment), tzinfo=ZoneInfo(alert.tz))
+            sendat_tz = datetime.combine(fields.Date.context_today(alert, fields.Datetime.now()), float_to_time(alert.notification_time), tzinfo=ZoneInfo(alert.tz))
             cron = alert.cron_id.sudo()
             lc = cron.lastcall
             if ((

--- a/addons/lunch/models/lunch_supplier.py
+++ b/addons/lunch/models/lunch_supplier.py
@@ -1,35 +1,22 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-import math
 from collections import defaultdict
-from datetime import datetime, time, timedelta, UTC
+from datetime import datetime, timedelta, UTC
 from textwrap import dedent
 from zoneinfo import ZoneInfo
 
 from odoo import api, fields, models, _
 from odoo.exceptions import UserError
 from odoo.fields import Domain
-from odoo.tools import float_round
+from odoo.tools.date_utils import float_to_time
 from odoo.tools.translate import LazyTranslate
 
 from odoo.addons.base.models.res_partner import _tz_get
 
 WEEKDAY_TO_NAME = ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun']
-CRON_DEPENDS = {'name', 'active', 'send_by', 'automatic_email_time', 'moment', 'tz'}
+CRON_DEPENDS = {'name', 'active', 'send_by', 'automatic_email_time', 'tz'}
 _lt = LazyTranslate(__name__)
 
-def float_to_time(hours, moment='am'):
-    """ Convert a number of hours into a time object. """
-    if hours == 12.0 and moment == 'pm':
-        return time.max
-    fractional, integral = math.modf(hours)
-    if moment == 'pm':
-        integral += 12
-    return time(int(integral), int(float_round(60 * fractional, precision_digits=0)), 0)
-
-def time_to_float(t):
-    return float_round(t.hour + t.minute/60 + t.second/3600, precision_digits=2)
-
 
 class LunchSupplier(models.Model):
     _name = 'lunch.supplier'
@@ -81,11 +68,6 @@ class LunchSupplier(models.Model):
 
     active = fields.Boolean(default=True)
 
-    moment = fields.Selection([
-        ('am', 'AM'),
-        ('pm', 'PM'),
-    ], default='am', required=True)
-
     delivery = fields.Selection([
         ('delivery', 'Delivery'),
         ('no_delivery', 'No Delivery')
@@ -114,8 +96,8 @@ class LunchSupplier(models.Model):
     show_confirm_button = fields.Boolean(compute='_compute_buttons')
 
     _automatic_email_time_range = models.Constraint(
-        'CHECK(automatic_email_time >= 0 AND automatic_email_time <= 12)',
-        'Automatic Email Sending Time should be between 0 and 12',
+        'CHECK(automatic_email_time >= 0 AND automatic_email_time <= 24)',
+        'Automatic Email Sending Time should be between 0 and 24',
     )
 
     @api.depends('phone')
@@ -130,7 +112,7 @@ class LunchSupplier(models.Model):
         for supplier in self:
             supplier = supplier.with_context(tz=supplier.tz)
 
-            sendat_tz = datetime.combine(fields.Date.context_today(supplier), float_to_time(supplier.automatic_email_time, supplier.moment), tzinfo=ZoneInfo(supplier.tz))
+            sendat_tz = datetime.combine(fields.Date.context_today(supplier), float_to_time(supplier.automatic_email_time), tzinfo=ZoneInfo(supplier.tz))
             cron = supplier.cron_id.sudo()
             lc = cron.lastcall
             if ((
@@ -314,7 +296,7 @@ class LunchSupplier(models.Model):
         for supplier in self:
             if supplier.send_by == 'mail':
                 now = now.astimezone(ZoneInfo(supplier.tz))
-                email_time = datetime.combine(fields.Date.context_today(supplier), float_to_time(supplier.automatic_email_time, supplier.moment), tzinfo=ZoneInfo(supplier.tz))
+                email_time = datetime.combine(fields.Date.context_today(supplier), float_to_time(supplier.automatic_email_time), tzinfo=ZoneInfo(supplier.tz))
                 supplier.order_deadline_passed = supplier.available_today and now > email_time
             else:
                 supplier.order_deadline_passed = not supplier.available_today
